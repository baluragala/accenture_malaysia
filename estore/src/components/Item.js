import React from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";
class Item extends React.PureComponent {
  constructor(props) {
    super(props);
    console.log("constructor");
  }

  // shouldComponentUpdate = (nextProps, nextState) => {
  //   console.log("will render???", this.props.product !== nextProps.product);
  //   return this.props.product !== nextProps.product;
  // };

  componentDidMount() {
    console.log("componentDidMount");
  }

  componentDidUpdate(prevProps, prevState) {
    console.log("componentDidUpdate test");
  }

  componentWillUnmount() {
    console.log("componentWillUnmount");
  }

  render() {
    let {
      product: { title, stock, price, id },
      onSell
    } = this.props;
    console.log("render", id);
    return (
      <div>
        <h3>
          <Link to={`/products/${id}`}>
            {title.toUpperCase()}({stock})
          </Link>
        </h3>
        <p>{price}</p>
        <button onClick={() => onSell(id)}>Sell</button>
      </div>
    );
  }
}

Item.defaultProps = {
  product: { id: 0, title: "N/A", price: -1, stock: -1 },
  onSell: () => {}
};

Item.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.number.isRequired,
    title: PropTypes.string.isRequired,
    stock: PropTypes.number.isRequired,
    price: PropTypes.number.isRequired
  }),
  onSell: PropTypes.func.isRequired
};

export default Item;
