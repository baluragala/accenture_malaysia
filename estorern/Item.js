import React from "react";
import { View, Text, Image, StyleSheet, Alert } from "react-native";
import { FontAwesome } from "@expo/vector-icons";

class Item extends React.Component {
  render() {
    let { title, addintionalInfo, price, rating, id } = this.props.product;
    return (
      <View style={{ height: 175 }}>
        <View style={[styles.container]}>
          <Image
            style={styles.image}
            source={{ uri: "https://dummyimage.com/100X150/000/fff" }}
          />
          <View style={[styles.infoContainer]}>
            <Text style={styles.title}>
              {id}.{title}
            </Text>
            <Text style={styles.addintionalInfo}>{addintionalInfo}</Text>
            <View style={[styles.ratingContainer]}>
              <Text>{rating}</Text>
              <FontAwesome name="star" size={10} color="white" />
            </View>
            <Text style={styles.price}>{price}</Text>
          </View>
          <FontAwesome
            name="heart-o"
            size={32}
            color="green"
            onPress={() => {
              Alert.alert("Favorite", "Would like to set this as favorite", [
                {
                  text: "Ok",
                  onPress: () => {
                    console.log("Ok");
                  }
                },
                {
                  text: "Cancel",
                  onPress: () => {
                    console.log("Cancel");
                  }
                }
              ]);
            }}
          />
        </View>
      </View>
    );
  }
}

export default Item;

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: "row",
    padding: 10,
    justifyContent: "space-between",
    borderColor: "grey",
    borderBottomWidth: StyleSheet.hairlineWidth
  },
  image: {
    width: 100,
    height: 150
  },
  infoContainer: {
    flex: 8,

    marginLeft: 10,
    marginRight: 10,
    padding: 5
  },
  title: {
    fontSize: 16,
    color: "blue",
    fontWeight: "bold"
  },
  ratingContainer: {
    flexDirection: "row",
    color: "white",
    borderWidth: 1,
    borderColor: "black",
    width: 40,
    justifyContent: "center",
    alignItems: "center",
    backgroundColor: "lightgreen"
  }
});
