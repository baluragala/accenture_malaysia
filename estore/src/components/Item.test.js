import React from "react";
import Item from "./Item";
import { configure, shallow } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { Link } from "react-router-dom";
import { shallowToJson } from "enzyme-to-json";

configure({ adapter: new Adapter() });

describe("Item component", function() {
  let p, wrapper;
  beforeEach(() => {
    p = { id: 1, title: "test", price: 10, stock: 10 };
    wrapper = shallow(<Item product={p} onSell={() => {}} />);
  });

  afterEach(() => {
    p = null;
    wrapper = null;
  });

  it("should render a product", function() {
    expect(wrapper).toBeTruthy();
  });

  it("should render a Link", function() {
    let linkWrapper = wrapper.find(Link);
    expect(linkWrapper.props().to).toBe("/products/1");
  });

  it("should invoke parents onSell", function() {
    let mockParentFn = jest.fn();
    wrapper = shallow(<Item product={p} onSell={mockParentFn} />);
    let buttonWrapper = wrapper.find("button");
    buttonWrapper.simulate("click");
    expect(mockParentFn).toHaveBeenCalledWith(1);
  });

  it("snapshots must be same", function() {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
