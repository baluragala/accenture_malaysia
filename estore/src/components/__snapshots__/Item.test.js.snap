// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item component snapshots must be same 1`] = `
<div>
  <h3>
    <Link
      replace={false}
      to="/products/1"
    >
      TEST
      (
      10
      )
    </Link>
  </h3>
  <p>
    10
  </p>
  <button
    onClick={[Function]}
  >
    Sell
  </button>
</div>
`;
